spannerlib/wrappers/spannerlib-ruby/lib/spannerlib/ffi.rb

```diff
@@ -0,0 +1,224 @@
+require 'rubygems'
+require 'bundler/setup'
+
+require 'google/protobuf'
+require 'google/rpc/status_pb'
+
+require 'ffi'
+
+module SpannerLib
+  extend FFI::Library
+  ffi_lib File.expand_path('../../shared/spannerlib.so', __dir__)
+
+  class GoString < FFI::Struct
+    layout :p,   :pointer,
+           :len, :long
+  end
+
+  # GoBytes is the Ruby representation of a Go byte slice 
+  class GoBytes < FFI::Struct
+    layout :p,   :pointer,
+           :len, :long,
+           :cap, :long
+  end
+
+  # Message is the common return type for all native functions.
+  class Message < FFI::Struct
+    layout :pinner,   :long_long, 
+           :code,     :int,       
+           :objectId, :long_long,  
+           :length,   :int,       
+           :pointer,  :pointer     
+  end
+
+  # --- Native Function Signatures ---
+  attach_function :CreatePool, [GoString.by_value], Message.by_value
+  attach_function :ClosePool, [:int64], Message.by_value
+  attach_function :CreateConnection, [:int64], Message.by_value
+  attach_function :CloseConnection, [:int64, :int64], Message.by_value
+  attach_function :WriteMutations, [:int64, :int64, GoBytes.by_value], Message.by_value
+  attach_function :BeginTransaction, [:int64, :int64, GoBytes.by_value], Message.by_value
+  attach_function :Commit, [:int64, :int64], Message.by_value
+  attach_function :Rollback, [:int64, :int64], Message.by_value
+  attach_function :Execute, [:int64, :int64, GoBytes.by_value], Message.by_value
+  attach_function :ExecuteBatch, [:int64, :int64, GoBytes.by_value], Message.by_value
+  attach_function :Metadata, [:int64, :int64, :int64], Message.by_value
+  attach_function :Next, [:int64, :int64, :int64, :int32, :int32], Message.by_value
+  attach_function :ResultSetStats, [:int64, :int64, :int64], Message.by_value
+  attach_function :CloseRows, [:int64, :int64, :int64], Message.by_value
+  attach_function :Release, [:int64], :void
+
+  # --- Ruby-friendly Wrappers ---
+
+  def self.create_pool(dsn)
+    dsn_str = dsn.to_s.dup
+    dsn_ptr = FFI::MemoryPointer.from_string(dsn_str)
+    
+    go_dsn = GoString.new
+    go_dsn[:p] = dsn_ptr
+    go_dsn[:len] = dsn_str.bytesize
+
+    message = CreatePool(go_dsn)
+    handle_object_id_response(message, "CreatePool")
+  end
+
+  def self.close_pool(pool_id)
+    message = ClosePool(pool_id)
+    handle_status_response(message, "ClosePool")
```

Comment:
nit: would it maybe make sense to let `handle_status_response` return `nil` when it is successful? Then you don't need to repeat the `nil` result in each function that calls `handle_status_response`.